### ISSUE TITLE
Are you confident the BlocProvider will work as intended?

### ISSUE THREAD

**Issue by Soundbytes** [felangel/bloc#20]
Looking at the `BlocProvider` code I found that you are using the `BuildContext.ancestorWidgetOfExactType` method instead of 
`BuildContext.inheritFromWidgetOfExactType`
From what I could gather this will prevent Widgets that use the `BlocProvider` from rebuilding when the Bloc is swapped for another one.
I have already created a version that derives `BlocProvider` from `StatelessWidget` because I don't need the addditional overhead that comes with using `InheritedWidget`. 
I haven't  made a pull request so far as I have no time to make the required tests to support my claim. Let me know if you are interested.

**Comment by felangel** [comment]
Hi @Soundbytes to clarify, the use case you're describing is when a BlocProvider's bloc has changed, the widget subtree should rebuild correct?

Also, can you please elaborate a bit on what you mean by not needing additional overhead from `InheritedWidget`?

It would also be super useful if you can include an example of your particular use case.
Thanks!

**Comment by Soundbytes** [comment]
Yep that is the use case and that is what InheritedWidget was made for. 
As I said I don't think that it will work at all with the BlocProvider because it requires to use a particular BuildContext method that makes some provisions behind the scenes to enable the behavior
[inheritFromWidgetOfExactType](https://docs.flutter.io/flutter/widgets/BuildContext/inheritFromWidgetOfExactType.html)
since it won't work anyway with 
[ancestorWidgetOfExactType](https://docs.flutter.io/flutter/widgets/BuildContext
/ancestorWidgetOfExactType.html)

Read this brilliant article for more information on the overhead introduced by InheritedWidget in comparison to a plain StatelessWidget: [Reactive Programming — Streams — BLoC](https://medium.com/flutter-community/reactive-programming-streams-bloc-6f0d2bd2d248)
(scroll down to the **Why not using an InheritedWidget?** paragraph) 

Here is the modified code. It seems to work fine. I am afraid I had no time to write tests. At least your tests did all pass.
```
class BlocProvider<T extends Bloc<dynamic, dynamic>> extends StatelessWidget {
  const BlocProvider({
    Key key,
    @required this.bloc,
    @required this.child,
  })  : assert(bloc != null),
        super(key: key);

  /// The Bloc which is to be made available throughout the subtree
  final T bloc;

  final Widget child;

  /// Method that allows widgets to access the bloc as long as their `BuildContext`
  /// contains a `BlocProvider` instance.
  static B of<B extends Bloc<dynamic, dynamic>>(BuildContext context) {
    Type typeOf<B>() => B;
    BlocProvider<B> provider =
        context.ancestorWidgetOfExactType(typeOf<BlocProvider<B>>());
    return provider.bloc;
  }

  @override
  Widget build(BuildContext context) => child;
}
```

**Comment by Soundbytes** [comment]
Opps, pressed the wrong button.

**Comment by felangel** [comment]
@Soundbytes so I completely agree that I the current implementation should have been using `inheritFromWidgetOfExactType` instead of `ancestorWidgetOfExactType`; however, I'm not quite sure I understand why the `StatelessWidget` implementation will perform better than the `InheritedWidget` implementation with `inheritFromWidgetOfExactType`.

Take the case where we have a really deep widget tree and we are calling `BlocProvider.of(context)`:

Scenario A. If the implementation is using `StatelessWidget` and `ancestorWidgetOfExactType`, this operation will take O(n).
Scenario B. If the implementation is using `InheritedWidget` and `inheritFromWidgetOfExactType`, this operation will take O(1).

I understand that in scenario B, the widget will now be rebuilt more often but I'm not confident that A is  definitively more performant than B. Thoughts?

**Comment by lukepighetti** [comment]
Could you please provide more information on your performance testing? Thanks!

**Comment by felangel** [comment]
sure I’ll publish the performance tests I ran in a couple of hours.

The basic setup was an app with a deep widget tree and a BlocProvider at the root. Along the way down the sub tree I made every widget request for the bloc via the BuildContext and then looked at TimelineSummary for various performance statistics.

I did several trials with each of the three setups:
InheritedWidget using ancestorWidgetOfExactType

InheritedWidget using inheritFromWidgetOfExactType

StatelessWidget using ancestorWidgetOfExactType
